service/src/main/kotlin/fi/espoo/evaka/vasu/VasuController.kt

```diff
@@ -88,32 +87,59 @@ class VasuController(
     fun getDocument(
         db: Database,
         user: AuthenticatedUser,
         @PathVariable id: VasuDocumentId
-    ): GetVasuDocumentResponse {
+    ): VasuDocument {
         Audit.VasuDocumentRead.log(id)
         accessControl.requirePermissionFor(user, Action.VasuDocument.READ, id)
 
         return db.connect { dbc ->
             dbc.read { tx ->
                 val doc = tx.getVasuDocumentMaster(id) ?: throw NotFound("template $id not found")
-                val entryIds = doc.content.followupEntries().map { doc.id to it.id }.toList()
-                val permittedActions =
-                    accessControl.getPermittedActions<VasuDocumentFollowupEntryId, Action.VasuDocumentFollowup>(
-                        tx,
-                        user,
-                        entryIds
+
+                val employeeIds = doc.content.sections.flatMap { section ->
+                    section.questions.flatMap { question ->
+                        if (question.type === VasuQuestionType.FOLLOWUP) {
+                            val followup = question as VasuQuestion.Followup
+                            followup.value.flatMap { setOfNotNull(it.authorId, it.edited?.editorId) }
+                        } else {
+                            emptySet()
+                        }
+                    }
+                }
+
+                val employeeNames = if (employeeIds.isEmpty()) emptyMap() else tx.getEmployeeNamesByIds(employeeIds)
+
+                doc.copy(
+                    content = doc.content.copy(
+                        sections = doc.content.sections.map { section ->
+                            section.copy(
+                                questions = section.questions.map { question ->
+                                    if (question.type === VasuQuestionType.FOLLOWUP) {
+                                        val followup = question as VasuQuestion.Followup
+
+                                        followup.copy(
+                                            value = followup.value.map { entry ->
+                                                entry.copy(
+                                                    authorName = entry.authorId?.let { employeeNames[it] },
+                                                    edited = entry.edited?.let { edited ->
+                                                        edited.copy(
+                                                            editorName = edited.editorId?.let { employeeNames[it] }
+                                                        )
+                                                    }
+                                                )
+                                            }
+                                        )
+                                    } else {
+                                        question
+                                    }
+                                }
+                            )
+                        }
```

Comment:
I think it wouldn't hurt to use named functions or variables here to add more semantics for better readability